Kernel/UBSanitizer.cpp

```diff
@@ -53,73 +58,156 @@ void __ubsan_handle_nonnull_arg(const NonnullArgData& data)
+void __ubsan_handle_out_of_bounds(const OutOfBoundsData& data, ValueHandle)
 {
     dbgln("KUBSAN: out of bounds access into array of {} ({}-bit), index type {} ({}-bit)", data.array_type.name(), data.array_type.bit_width(), data.index_type.name(), data.index_type.bit_width());
     print_location(data.location);
 }
 
-void __ubsan_handle_type_mismatch_v1(const TypeMismatchData&, void*);
-void __ubsan_handle_type_mismatch_v1(const TypeMismatchData& data, void*)
+void __ubsan_handle_type_mismatch_v1(const TypeMismatchData&, ValueHandle);
+void __ubsan_handle_type_mismatch_v1(const TypeMismatchData& data, ValueHandle)
 {
     dbgln("KUBSAN: type mismatch, {} ({}-bit)", data.type.name(), data.type.bit_width());
     print_location(data.location);
 }
+
+// FIXME: Causes a triple fault on boot
+void __ubsan_handle_alignment_assumption(const AlignmentAssumptionData&, ValueHandle, ValueHandle, ValueHandle);
+void __ubsan_handle_alignment_assumption(const AlignmentAssumptionData& data, ValueHandle pointer, ValueHandle alignment, ValueHandle offset)
+{
+    if (offset) {
+        dbgln(
+            "KUBSAN: assumption of {:p} byte alignment (with offset of {:p} byte) for pointer {:p}"
+            "of type {} failed",
+            alignment, offset, pointer, data.type.name());
+    } else {
+        dbgln("KUBSAN: assumption of {:p} byte alignment for pointer {:p}"
+              "of type {} failed",
+            alignment, pointer, data.type.name());
+    }
+    // dbgln("KUBSAN: Assumption of pointer allignment failed");
+    print_location(data.location);
+}
+
+void __ubsan_handle_builtin_unreachable(const UnreachableData&);
+void __ubsan_handle_builtin_unreachable(const UnreachableData& data)
+{
+    dbgln("KUBSAN: execution reached an unreachable program point");
+    print_location(data.location);
+}
+
+void __ubsan_handle_missing_return(const UnreachableData&);
+void __ubsan_handle_missing_return(const UnreachableData& data)
+{
+    dbgln("KUBSAN: execution reached the end of a value-returning function without returning a value");
+    print_location(data.location);
+}
+
+void __ubsan_handle_implicit_conversion(const ImplicitConversionData&, ValueHandle, ValueHandle);
+void __ubsan_handle_implicit_conversion(const ImplicitConversionData& data, ValueHandle, ValueHandle)
+{
+    const char* src_signed = data.from_type.is_signed() ? "" : "un";
+    const char* dst_signed = data.to_type.is_signed() ? "" : "un";
+    dbgln("KUBSAN: implicit conversion from type {} ({}-bit, {}signed) to type {} ({}-bit, {}signed)",
+        data.from_type.name(), data.from_type.bit_width(), src_signed, data.to_type.name(), data.to_type.bit_width(), dst_signed);
+    print_location(data.location);
+}
+
+void __ubsan_handle_invalid_builtin(const InvalidBuiltinData);
+void __ubsan_handle_invalid_builtin(const InvalidBuiltinData data)
+{
+    dbgln("KUBSAN: passing invalid argument");
+    print_location(data.location);
+}
+
+// FIXME: Causes a triple fault on boot
```

Comment:
Is it a trouble fault? I tried adding this previously and when I stepped through in gdb it looked like it was faulting on the `movl %ebx, multiboot_info_ptr` instruction before we even enter any C++ code? 
``` asm
    /* jump into C++ land */
    addl $0xc0000000, %ebx
    movl %ebx, multiboot_info_ptr


    call init
    add $4, %esp
```

Reply:
Im not using any debug tools and the only output I get is the following (similar to the one mentioned above):

<details>
<summary>Build and Error Log</summary>

```
> Executing task: wsl bash -c "cd Build && ninja install && ninja image" <

[0/2] Re-checking globbed directories...
[1/2] Re-running CMake...
        kernel fsanitize flag:shift
        kernel fsanitize flag:shift-exponent
        kernel fsanitize flag:shift-base
        kernel fsanitize flag:integer-divide-by-zero
        kernel fsanitize flag:unreachable
        kernel fsanitize flag:vla-bound
        kernel fsanitize flag:return
        kernel fsanitize flag:signed-integer-overflow
        kernel fsanitize flag:bounds
        kernel fsanitize flag:bounds-strict
        kernel fsanitize flag:object-size
        kernel fsanitize flag:float-divide-by-zero
        kernel fsanitize flag:float-cast-overflow
        kernel fsanitize flag:nonnull-attribute
        kernel fsanitize flag:returns-nonnull-attribute
        kernel fsanitize flag:bool
        kernel fsanitize flag:enum
        kernel fsanitize flag:vptr
        kernel fsanitize flag:builtin
        kernel fsanitize flag:pointer-overflow
CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon:
  /mnt/serenity/Base/res/icons/32x32/audio-volume-high.png
Call Stack (most recent call first):
  Userland/MenuApplets/Audio/CMakeLists.txt:5 (serenity_app)


CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon: /mnt/serenity/Base/res/icons/32x32/clipboard.png
Call Stack (most recent call first):
  Userland/MenuApplets/ClipboardHistory/CMakeLists.txt:6 (serenity_app)


CMake Warning at Meta/CMake/utils.cmake:89 (message):
  Missing medium app icon: /mnt/serenity/Base/res/icons/32x32/network.png
Call Stack (most recent call first):
  Userland/MenuApplets/Network/CMakeLists.txt:5 (serenity_app)


-- Configuring done
-- Generating done
-- Build files have been written to: /mnt/serenity/Build
[0/2] Re-checking globbed directories...
[240/241] Install the project...
-- Install configuration: ""
[0/2] Re-checking globbed directories...
[0/1] cd /mnt/serenity/Build && /usr/bin/cmake -E env SERENITY_ROOT=/mnt/serenity SERENITY_ARCH=i686 /mnt/serenity/Meta/build-image-qemu.sh
setting up disk image...
Formatting '_disk_image', fmt=raw size=222298112
done
creating new filesystem... done
mounting filesystem... mount: mnt/: mount failed: Operation not permitted.
mount failed but genext2fs exists, use it instead
done
installing base system... done
creating initial filesystem structure... done
creating utmp file... done
setting up device nodes folder... done
writing version file... done
installing users... done
adding some desktop icons...done
installing shortcuts... done
installing 'checksum' variants... done
done

Terminal will be reused by tasks, press any key to close it.

> Executing task: qemu-system-i386 -L "C:\Program Files\qemu" -s -m 4G -cpu max -d cpu_reset,guest_errors -no-reboot -smp 2 -device VGA,vgamem_mb=64 -drive file=_disk_image,format=raw,index=0,media=disk -device ich9-ahci -usb -debugcon stdio -soundhw pcspk -device sb16 -netdev user,id=breh,hostfwd=tcp:127.0.0.1:8888-10.0.2.15:8888,hostfwd=tcp:127.0.0.1:8823-10.0.2.15:23 -device e1000,netdev=breh -kernel Kernel/Kernel -append hello <

CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000 
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000000
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=00000000 EFL=00000000 [-------] CPL=0 II=0 A20=0 SMM=0 HLT=0
ES =0000 00000000 00000000 00000000
CS =0000 00000000 00000000 00000000
SS =0000 00000000 00000000 00000000
DS =0000 00000000 00000000 00000000
FS =0000 00000000 00000000 00000000
GS =0000 00000000 00000000 00000000
LDT=0000 00000000 00000000 00000000
TR =0000 00000000 00000000 00000000
GDT=     00000000 00000000
IDT=     00000000 00000000
CR0=00000000 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=00000000 DR7=00000000
CCS=00000000 CCD=00000000 CCO=DYNAMIC 
EFER=0000000000000000
FCW=0000 FSW=0000 [ST=0] FTW=ff MXCSR=00000000
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
qemu-system-i386: multiboot knows VBE. we don't
qemu-system-i386: warning: '-soundhw pcspk' is deprecated, please set a backend using '-machine pcspk-audiodev=<name>' instead
CPU Reset (CPU 0)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=1
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=DYNAMIC 
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
CPU Reset (CPU 1)
EAX=00000000 EBX=00000000 ECX=00000000 EDX=00000663
ESI=00000000 EDI=00000000 EBP=00000000 ESP=00000000
EIP=0000fff0 EFL=00000002 [-------] CPL=0 II=0 A20=1 SMM=0 HLT=0
ES =0000 00000000 0000ffff 00009300
CS =f000 ffff0000 0000ffff 00009b00
SS =0000 00000000 0000ffff 00009300
DS =0000 00000000 0000ffff 00009300
FS =0000 00000000 0000ffff 00009300
GS =0000 00000000 0000ffff 00009300
LDT=0000 00000000 0000ffff 00008200
TR =0000 00000000 0000ffff 00008b00
GDT=     00000000 0000ffff
IDT=     00000000 0000ffff
CR0=60000010 CR2=00000000 CR3=00000000 CR4=00000000
DR0=00000000 DR1=00000000 DR2=00000000 DR3=00000000
DR6=ffff0ff0 DR7=00000400
CCS=00000000 CCD=00000000 CCO=EFLAGS  
EFER=0000000000000000
FCW=037f FSW=0000 [ST=0] FTW=00 MXCSR=00001f80
FPR0=0000000000000000 0000 FPR1=0000000000000000 0000
FPR2=0000000000000000 0000 FPR3=0000000000000000 0000
FPR4=0000000000000000 0000 FPR5=0000000000000000 0000
FPR6=0000000000000000 0000 FPR7=0000000000000000 0000
XMM00=00000000000000000000000000000000 XMM01=00000000000000000000000000000000
XMM02=00000000000000000000000000000000 XMM03=00000000000000000000000000000000
XMM04=00000000000000000000000000000000 XMM05=00000000000000000000000000000000
XMM06=00000000000000000000000000000000 XMM07=00000000000000000000000000000000
Triple fault
```

</details>

As you can see the last thing qemu tells me is that there was a tripple fault and the proccessor state at that time.  
If I'm reading it correctly the instruction pointer is 0x0000fff0 which doesnt look that good.

You get pretty much the same stacktrace when trying any of the other two failing flags.  

As you said the fault probably happens before any C++ generated code is run, maybe that is the problem, that gcc tries to audit something before this level of auditing is even possible, like trying to call the handlers without even having a stack or in the wrong bit-mode.

Reply:
Is it possible, that ebx is 0 at that point in time and we add 0xc0000000 to 0?  
I havent found any other mention of ebx in that file.

My guess would there fore be:
```asm
/* jump into C++ land */
    movl $0xc0000000, %ebx
    movl %ebx, multiboot_info_ptr


    call init
    add $4, %esp
```

but with that I still get the same error

Reply:
After stepping through with gdb I can say with confidence, that it does not crash at this point in time

Reply:
Oh ive misread your question, you did not mean the pointer-overflow flag but the allignment flag, and then yes it does pagefault (I really dont know how, but the error code I see is 0xe wich is a pagefault) there, but there arent any faulthandlers regustered at that point therefore the fault escalates to a triple fault